federation/pkg/federation-controller/sync/controller.go

```diff
@@ -424,29 +475,54 @@ type clusterObjectAccessorFunc func(clusterName string) (interface{}, bool, erro
 // clusterOperations returns the list of operations needed to synchronize the state of the given object to the provided clusters
 func clusterOperations(adapter federatedtypes.FederatedTypeAdapter, clusters []*federationapi.Cluster, obj pkgruntime.Object, accessor clusterObjectAccessorFunc) ([]util.FederatedOperation, error) {
 	key := federatedtypes.ObjectKey(adapter, obj)
-	operations := make([]util.FederatedOperation, 0)
+	current := make(map[string]pkgruntime.Object)
 	for _, cluster := range clusters {
 		clusterObj, found, err := accessor(cluster.Name)
 		if err != nil {
 			return nil, fmt.Errorf("Failed to get %s %q from cluster %q: %v", adapter.Kind(), key, cluster.Name, err)
 		}
-		// The data should not be modified.
-		desiredObj := adapter.Copy(obj)
 
-		var operationType util.FederatedOperationType = ""
 		if found {
-			clusterObj := clusterObj.(pkgruntime.Object)
-			if !adapter.Equivalent(desiredObj, clusterObj) {
+			current[cluster.Name] = clusterObj.(pkgruntime.Object)
+		} else {
+			current[cluster.Name] = nil
+		}
+	}
+
+	scheduled := make(map[string]pkgruntime.Object)
+	if h := adapter.GetScheduleHook(); h != nil {
+		var err error
+		// The hook implementation should ensure not to mutate the passed objects.
+		// Its recommended that implementation also keeps exactly same number of entries
+		// as current.
+		scheduled, err = h(obj, current)
+		if err != nil {
+			return nil, fmt.Errorf("Failure executing Schedule Hook on %s %q: %v", adapter.Kind(), key, err)
+		}
+	} else {
+		for clusterKey := range current {
+			// The same object propagated into all clusters.
+			desiredObj := adapter.Copy(obj)
+			scheduled[clusterKey] = desiredObj
+		}
+
+	}
+
+	operations := make([]util.FederatedOperation, 0)
+	for cluster, currentObj := range current {
```

Comment:
My point is that it seems simpler to have only one map in play, since scheduled is populated in both cases.  Having 2 maps seems like accidental complexity to me.